### DISCUSSION TITLE
Configuring/disabling subscribe limiter/throttle

### DISCUSSION THREAD

**Discussion by gaxx** [emqx/emqx#8241] [Q&A]
Hello!

EMQX version: 5.0-RC4

I do have a use case where a client needs to subscribe ~8-10k (retained) topics at once. It would be good to get those messages in few seconds but currently 8k subscribes takes almost a minute to deliver. Additionally it requires some tweaking to get messages in less than a minute (not subscribing all at once but sending subscriptions in smaller batches with a delay between each). Without "tweaks" client disconnects/reconnects several times and it will take even longer.

My testing shows that there is a limiter/throttle for handling those subscriptions since if I use a single subscription like my/test/topics/#, messages are delivered very fast (10k msg/sec, I believe that my python script is the bottle-neck in that case) even if there are many times more messages to deliver.

When a new client connects it can reach about 1000 subscription handles per second (of retained topics) if it does only small amount of subscriptions (like 500 at once). If it does it again, the rate will drop down (~150 seems to be the bottom). If I wait a while and subscribe 500 topics again, the rate is something like 600 subs/second (so it never seem to get back to 1000 sub/second rate).

Anyway, I didn't succeed to find any configuration that would affect to this rate/limit. Is there one (or multiple)? If not, is it possible to get one or is there some technical reason not to provide one?

**Comment by lafirest** [comment]
@gaxx 
Thanks for your input
In 5.0, reatiner does not have any rate limit by default, but this can easily cause the client to receive an excessively large number of messages and then crash, so we introduced a rate limit system for retainer. 
But this doesn't seem to be the same problem you encountered, do you expect to get all retainer messages at once, and as fast as possible?

**Reply by gaxx** [reply]
@lafirest
Thanks for quick answer!

Yes, I am expecting to get those messages as fast as possible. In my case I have a topic tree like data/devices/my_device_3/<sub_topic_tree> (where sub_topic_tree contains ~40k different topics).
If I subscribe those topics with single subscription like data/devices/my_device_3/# it will return those messages very fast. But if I send 10k different subscriptions (to pick only relevant topics that I want to monitor also in the future) this limiter/throttle will affect. It is not a problem if I have only like 1k different subscriptions but for example with 5k subscription handling becomes quite slow.

P.S. did you mean that 4.x doesn't have rate limit and you introduced one in 5.0?

**Reply by zmstone** [reply]
Hi @gaxx 

Inside EMQX, there is one-process for each MQTT client, this single process (fairly scheduled among others) is handling all the subscribe requests, as well as querying(or scanning in some unlucky cases) the retained message db-table.
Subscribing to one wildcard subscription allows querying (or scanning in some unlucky cases) in one go.
Subscribing to 10K topics the poor process will have to query (or scan in some unlucky cases) 10K times.
I believe this overhead is the main reason why you are seeing the slowdown.

As for the re-subscription being slower than original, I don't have a straight answer without having a closer look into the code base.

Arguably, one single subscriber subscribing to 10K different topics isn't really a good practice.
MQTT's topic structure encourages the use of wildcards so there is no need for large amount of subscribes.
Maybe you can re-org a bit on the topics so they can be better partitioned by deeper level topic words(segments)?

**Reply by lafirest** [reply]
@gaxx 
<img width="636" alt="image" src="https://user-images.githubusercontent.com/22971499/174253705-3d839fc4-201d-4886-a3a4-5d869f880b00.png">
As you can see, I test with a 100k dataset, which has done in 1 second. But this is not safe, you must close or adjust the force shutdown strategy.
So there is no rate limit for retainer by default, the rate bottleneck depends on your client.

**Reply by gaxx** [reply]
@lafirest It is true that message delivery is very fast and it wouldn't be a problem if there is only small amount of subscribes (no matter how many retained topics those matches):

gaxx wrote:
> If I subscribe those topics with single subscription like data/devices/my_device_3/# it will return those messages very fast. But if I send 10k different subscriptions (to pick only relevant topics that I want to monitor also in the future) this limiter/throttle will affect.

lafirest wrote:
> So there is no rate limit for retainer by default, the rate bottleneck depends on your client.

My test client is not such bottleneck with other MQTT broker implementation we have tested (Mosquitto, VerneMQ).

@zmstone You are right that the way we are using subscriptions is not the most optimal way (although this has not been a problem with other MQTT broker implementations lately). We will try to find some easy ways to reduce number of subscriptions so that EMQX can handle it fast enough. I guess that those subscriptions are queued which causes this throttling like behavior.

Now I am just thinking loud: is it possible that long subscription queue is expensive to handle (maybe some inefficient algorithm there)? Or is it just a some locking mechanism that is used accessing retained message database...

**Reply by zmstone** [reply]
Hi @gaxx
Very good thinking. 
There is very likely quite some room for optimisation to large amount of per-client subscriptions.
e.g. handle in batch(es).
